engine/crates/engine-v2/src/response/write/mod.rs

```diff
@@ -368,6 +368,20 @@ impl ResponsePart {
             inner: Rc::new(RefCell::new(self)),
         }
     }
+
+    pub fn first_error_message(&self) -> Option<String> {
+        self.errors.first().map(|s| s.message.to_string())
+    }
+
+    /// True if this request was blocked by the query plan. Avoid logging any request details if this is true.
+    pub fn blocked_in_planning(&self) -> bool {
```

Comment:
What you want is this:

```
pub fn subgraph_errors(&self) -> impl Iterator<Item = &GraphqlError> + '_ {
  self.errors.iter().filter(|e| matches!(e.code ErrorCode::SubgraphError| ErrorCode::SubgraphInvalidResponseError |ErrorCode::SubgraphRequestError))
}
```

And then use the first message of this to decide what to do from it.